cdc/sink/dmlsink/mq/dmlproducer/pulsar_dml_producer.go

```diff
@@ -0,0 +1,312 @@
+	}
+
+	producer, err := p.GetProducerByTopic(topic)
+	if err != nil {
+		return err
+	}
+
+	// if for stress test record , add count to message callback function
+
+	producer.SendAsync(ctx, data,
+		func(id pulsar.MessageID, m *pulsar.ProducerMessage, err error) {
+			// fail
+			if err != nil {
+				e := cerror.WrapError(cerror.ErrPulsarAsyncSendMessage, err)
+				log.Error("Pulsar DML producer async send error",
+					zap.String("namespace", p.id.Namespace),
+					zap.String("changefeed", p.id.ID),
+					zap.Error(err))
+				mq.IncPublishedDMLFail(topic, p.id.ID, message.GetSchema())
+				p.errChan <- e
+			} else if message.Callback != nil {
+				// success
+				message.Callback()
+				mq.IncPublishedDMLSuccess(topic, p.id.ID, message.GetSchema())
+			}
+		})
+
+	mq.IncPublishedDMLCount(topic, p.id.ID, message.GetSchema())
+
+	return nil
+}
+
+func (p *pulsarDMLProducer) Close() {
+	// We have to hold the lock to synchronize closing with writing.
+	p.closedMu.Lock()
+	defer p.closedMu.Unlock()
+	// If the producer has already been closed, we should skip this close operation.
+	if p.closed {
+		// We need to guard against double closing the clients,
+		// which could lead to panic.
+		log.Warn("Pulsar DML producer already closed",
+			zap.String("namespace", p.id.Namespace),
+			zap.String("changefeed", p.id.ID))
+		return
+	}
+	close(p.failpointCh)
+	// Notify the run loop to exit.
+	close(p.closedChan)
+	p.closed = true
+
+	start := time.Now()
+	keys := p.producers.Keys()
+	for _, topic := range keys {
+		p.producers.Remove(topic) // callback func will be called
+		topicName, _ := topic.(string)
+		log.Info("Async client closed in pulsar "+
+			"DML producer", zap.Duration("duration", time.Since(start)),
+			zap.String("namespace", p.id.Namespace),
+			zap.String("changefeed", p.id.ID), zap.String("topic", topicName))
+	}
+
+	p.client.Close()
+}
+
+// newProducer creates a pulsar producer
+// One topic is used by one producer
+func newProducer(
+	pConfig *config.PulsarConfig,
+	client pulsar.Client,
+	topicName string,
+) (pulsar.Producer, error) {
+	po := pulsar.ProducerOptions{
+		Topic: topicName,
+	}
+	if pConfig.BatchingMaxMessages != nil {
+		po.BatchingMaxMessages = *pConfig.BatchingMaxMessages
+	}
+	if pConfig.BatchingMaxPublishDelay != nil {
+		po.BatchingMaxPublishDelay = pConfig.BatchingMaxPublishDelay.Duration()
+	}
+	if pConfig.CompressionType != nil {
+		po.CompressionType = pConfig.CompressionType.Value()
+		po.CompressionLevel = pulsar.Default
+	}
+	if pConfig.SendTimeout != nil {
+		po.SendTimeout = pConfig.SendTimeout.Duration()
+	}
+
+	producer, err := client.CreateProducer(po)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Info("new pulsar producer ok", zap.String("topic:", topicName))
+
+	return producer, nil
+}
+
+func (p *pulsarDMLProducer) getProducer(topic string) (pulsar.Producer, bool) {
+	target, ok := p.producers.Get(topic)
+	if ok {
+		producer, ok := target.(pulsar.Producer)
```

Comment:
Is it possible to fetch a target which is not `pulsar.Producer` ?

Reply:
No,it use to "GetProducerByTopic" to add and get from "getProducer" only .
The type is interface because lru set the key and value type are interface . I need to change it to pulsar.Producer